hazelcast/proxy/ringbuffer.py

```diff
@@ -157,7 +157,7 @@ def read_many(self, start_sequence, min_count, max_count):
         """
         check_not_negative(start_sequence, "sequence can't be smaller than 0")
         check_true(max_count >= min_count, "max count should be greater or equal to min count")
-        check_true(min_count <= self.capacity().result(), "min count should be smaller or equal to capacity")
+        #check_true(min_count <= self.capacity().result(), "min count should be smaller or equal to capacity")
```

Comment:
The correct validation for this line is:
```
check_true(max_count <= self.capacity().result(), "max count should be smaller or equal to capacity")
```

Reply:
@asimarslan,  @mdumandag  Sorry for this extreme delay. I was overloaded by my day job and the other open source development stuff. Now I'm back to the Hazelcast community.  

I changed the validation line and added some logic to call `read_many` in `_MessageListener._next` function. Without this modification, `read_many` call in the ReliableTopic implementation throws the exception:

```
HazelcastClient.LifecycleService: INFO: [dev] [hz.client_1] (20200531 - 2df48a6) HazelcastClient is CONNECTED
HazelcastClient: INFO: [dev] [hz.client_1] Client started.
HazelcastClient.AsyncoreReactor: ERROR: [dev] [hz.client_1] Error in Reactor Thread
Traceback (most recent call last):
  File "/home/buraksezer/Projects/python/hazelcast-python-client/hazelcast/reactor.py", line 46, in _loop
    self._check_timers()
  File "/home/buraksezer/Projects/python/hazelcast-python-client/hazelcast/reactor.py", line 66, in _check_timers
    if timer.check_timer(now):
  File "/home/buraksezer/Projects/python/hazelcast-python-client/hazelcast/reactor.py", line 277, in check_timer
    self.timer_ended_cb()
  File "/home/buraksezer/Projects/python/hazelcast-python-client/hazelcast/proxy/reliable_topic.py", line 211, in _read_many
    future = self._proxy.ringbuffer.read_many(self._sequence, 1, self._proxy.config.read_batch_size)
  File "/home/buraksezer/Projects/python/hazelcast-python-client/hazelcast/proxy/ringbuffer.py", line 160, in read_many
    check_true(max_count <= self.capacity().result(), "max count should be smaller or equal to capacity")
  File "/home/buraksezer/Projects/python/hazelcast-python-client/hazelcast/future.py", line 58, in result
    self._reactor_check()
  File "/home/buraksezer/Projects/python/hazelcast-python-client/hazelcast/future.py", line 70, in _reactor_check
    "Synchronous result for incomplete operation must not be called from Reactor thread. "
RuntimeError: Synchronous result for incomplete operation must not be called from Reactor thread. Use add_done_callback instead.
```

It seems that the corrected validation (max, instead of min) breaks some tests in `ringbuffer` implementation.

Reply:
No problem at all, we are happy to see you back.

I checked the line and the error. I also played with the version that has no capacity check but it seems there is no check on the server-side. So, when you have a ring buffer with 10 elements capacity, if you make a `read_many` request with `min_count` larger than that, the server never sends back a response (waits to get at least `min_count` items even if it is higher than its capacity).

So, I guess we have to have some kind of capacity check to fail on bad user requests. I checked your implementation but it seems that even if the capacity check fails, the future returned to the user is not resolved with that failure. I tweaked it a little bit to do that. What do you think about it? I did some tests and it seems it is working.

```python

        check_not_negative(start_sequence, "sequence can't be smaller than 0")
        check_true(max_count >= min_count, "max count should be greater or equal to min count")
        check_true(max_count < MAX_BATCH_SIZE, "max count can't be greater than %d" % MAX_BATCH_SIZE)

        future = Future()

        def set_result(result):
            try:
                future.set_result(result.result())
            except:
                future.set_exception(sys.exc_info()[1], sys.exc_info()[2])

        def check_capacity(capacity):
            try:
                capacity = capacity.result()
                check_true(min_count <= capacity, "min count: %d should be smaller or equal to capacity: %d"
                           % (min_count, capacity))
                f = self._encode_invoke(ringbuffer_read_many_codec, response_handler=self._read_many_response_handler,
                                        start_sequence=start_sequence, min_count=min_count,
                                        max_count=max_count, filter=None)
                f.add_done_callback(set_result)
            except:
                future.set_exception(sys.exc_info()[1], sys.exc_info()[2])

        self.capacity().add_done_callback(check_capacity)

        return future

```